README.md
# Laravel API Json Response

[![Latest Version on Packagist](https://img.shields.io/packagist/v/jornatf/laravel-api-json-response.svg?style=flat-square)](https://packagist.org/packages/jornatf/laravel-api-json-response)
[![GitHub Tests Action Status](https://img.shields.io/github/actions/workflow/status/jornatf/laravel-api-json-response/run-tests.yml?branch=main&label=tests&style=flat-square)](https://github.com/jornatf/laravel-api-json-response/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/actions/workflow/status/jornatf/laravel-api-json-response/fix-php-code-style-issues.yml?branch=main&label=code%20style&style=flat-square)](https://github.com/jornatf/laravel-api-json-response/actions?query=workflow%3A"Fix+PHP+code+style+issues"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/jornatf/laravel-api-json-response.svg?style=flat-square)](https://packagist.org/packages/jornatf/laravel-api-json-response)

A Laravel Package that returns a JSON Response for APIs. In some methods, you can return cool Json response for your API.

> #### If you like this package you can [Buy me a Coffee](https://www.buymeacoffee.com/jornatf) ☕️

## Installation

### Via composer:

```bash
composer require jornatf/laravel-api-json-response
```

## Usage

### Example 1:

> This example shows you how to use the basic required methods.

```php
<?php

namespace App\Http\Controllers;

use App\Models\Post;
use App\Http\Controllers\Controller;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Validator;

class PostController extends Controller
{
    public function store(Request $request)
    {
        $validator = Validator::make($request->all(), [
            'title' => 'required|unique:posts|max:255',
            'body' => 'required',
        ]);

        if ($validator->fails()) {
            // Status code 400 for "Bad Request"
            return ApiResponse::response(400)
                ->addDetails($validator->errors())
                ->json();
        }

        $post = Post::create($validator->validated());

        // Status code 200 for "Ok" or 201 for "Created"
        return ApiResponse::response(201)
            ->addDatas($post)
            ->json();
    }
}
```

**Success response:**

```json
{
    "success": {
        "status": 201,
        "message": "Created",
        "data": {
            "id": 1,
            "title": "Morbi in diam id dolor vehicula finibus",
            "content": "<p>Lorem ipsum dolor sit amet, ...</p>",
            "created_at": "2023-04-20 00:00:00",
            "updated_at": "2023-04-20 00:00:00"
        }
    }
}
```

**Error response:**

```json
{
    "error": {
        "status": 404,
        "message": "Not Found"
    }
}
```

### Exemple 2:

> This example shows you how to use a method to find a model and return a JSON Reponse in a single line of code.

```php
<?php

namespace App\Http\Controllers;

use App\Models\Post;
use App\Http\Controllers\Controller;

class PostController extends Controller
{
	public function show(int $post_id)
	{
		return ApiResponse::find(Post::class, $post_id)->json();
	}
}
```

**If model found:**

```json
{
    "success": {
        "status": 200,
        "message": "Post Found",
        "data": {
            "id": 1,
            "title": "Morbi in diam id dolor vehicula finibus",
            "content": "<p>Lorem ipsum dolor sit amet, ...</p>",
            "created_at": "2023-04-20 00:00:00",
            "updated_at": "2023-04-20 00:00:00"
        }
    }
}
```

**Else:**

```json
{
    "error": {
        "status": 404,
        "message": "Post Not Found"
    }
}
```

## Documentation

### Available methods:

```php
<?php

// First, you can instantiate response with a status code (required):
$response = ApiResponse::response(int $statusCode);
// or directly find a model by id:
$response = ApiResponse::find(Model::class, int $id);

// To add custom message:
$response->addMessage(string $message);

// To add datas to return when success:
$response->addDatas(array $datas);

// To add details (e.g. validator errors) when error:
$response->addDetails(mixed $details);
// or
$response->addDetails(array $details);
$response->addDetails(array $details);

// Last, formate response into json (required):
$response->json()
```

### Available status code:

| Status code | Response type | Default message              |
| ----------- | ------------- | ---------------------------- |
| `200`       | `success`     | Ok<br>Model Found            |
| `201`       | `success`     | Created                      |
| `202`       | `success`     | Accepted                     |
| `400`       | `error`       | Bad Request                  |
| `401`       | `error`       | Unauthorized                 |
| `403`       | `error`       | Forbidden                    |
| `404`       | `error`       | Not Found<br>Model Not Found |
| `405`       | `error`       | Method Not Allowed           |
| `408`       | `error`       | Request Timeout              |
| `429`       | `error`       | Too Many Requests            |
| `500`       | `error`       | Internal Server Error        |
| `502`       | `error`       | Bad Gateway                  |
| `503`       | `error`       | Service Unavailable          |

## Testing

```bash
composer test
```

## Changelog

> Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

> Feel free to contribute to this project to improve with new features or fix bugs 👍

## Credits

-   [Jordan](https://github.com/jornatf)
-   [All Contributors](../../contributors)

## License

The MIT License (MIT).

> Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to will be documented in this file.

## v1.1.0 - 2023-04-23

- improvements and new features
- updated README

Use a `find()` method to get a model and return a JSON Response in a single line like this:

```php
// Finds and returns User with id 1
return ApiResponse::find(User::class, 1)->json();

```
> See [README](README.md) for more details.

## v1.0.0 - 2023-04-22

Laravel API Json Response is a Laravel Package that returns a JSON Response for APIs. In some methods, you can return cool Json response for your API.


tests/Pest.php
<?php

use Jornatf\LaravelApiJsonResponse\Tests\TestCase;

uses(TestCase::class)->in(__DIR__);


tests/TestCase.php
<?php

namespace Jornatf\LaravelApiJsonResponse\Tests;

use Illuminate\Database\Eloquent\Factories\Factory;
use Jornatf\LaravelApiJsonResponse\ApiResponseServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

class TestCase extends Orchestra
{
    protected function setUp(): void
    {
        parent::setUp();

        Factory::guessFactoryNamesUsing(
            fn (string $modelName) => 'Jornatf\\LaravelApiJsonResponse\\Database\\Factories\\'.class_basename($modelName).'Factory'
        );
    }

    protected function getPackageProviders($app)
    {
        return [
            ApiResponseServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
        config()->set('database.default', 'testing');

        /*
        $migration = include __DIR__.'/../database/migrations/create_skeleton_table.php.stub';
        $migration->up();
        */
    }
}


tests/ApiResponseTest.php
<?php

use Jornatf\LaravelApiJsonResponse\ApiResponse;

/*
it('can test', function () {
    expect(true)->toBeTrue();
});
*/

it('returns json', function () {
    $response = (new ApiResponse)->response(200)
        ->json();

    expect($response->getContent())->toBeJson();
});

it('returns success json response (customized)', function () {
    $response = (new ApiResponse)->response(200)
        ->addMessage('Datas successfully retrieved')
        ->addDatas(['foo' => 'bar'])
        ->json();

    expect($response->getContent())->toBeJson();
});

it('returns error json response (customized)', function () {
    $response = (new ApiResponse)->response(404)
        ->addMessage('No data found for this request')
        ->addDetails(['id' => 9999])
        ->json();

    expect($response->getContent())->toBeJson();
});

it('throws exception: status code is not valid', function () {
    (new ApiResponse)->response(100)->json();
})->throws(Exception::class);

it('throws exception: status code is not an integer', function () {
    (new ApiResponse)->response('string')->json();
})->throws(Exception::class);

it('throws exception: datas is not an array', function () {
    (new ApiResponse)->response(200)->addDatas('not an array')->json();
})->throws(Exception::class);


src/Traits/HasJsonResponse.php
<?php

namespace Jornatf\LaravelApiJsonResponse\Traits;

trait HasJsonResponse
{
    /**
     * Success json response structure.
     *
     * @return \Illuminate\Http\Response
     */
    protected function successResponse(int $status, string $message, array $datas = [])
    {
        $response = [
            'success' => [
                'status' => $status,
                'message' => $message,
                'data' => $datas,
            ],
        ];

        return response()->json($response, $status);
    }

    /**
     * Error json response structure.
     *
     * @return \Illuminate\Http\Response
     */
    protected function errorResponse(int $status, string $message, array $details = [])
    {
        $response = [
            'error' => [
                'status' => $status,
                'message' => $message,
            ],
        ];

        foreach ($details as $detail) {
            $response['error']['details'][] = $detail;
        }

        return response()->json($response, $status);
    }

    /**
     * Success 200 : OK.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseOk(string $message, array $datas = [])
    {
        return $this->successResponse(200, $message, $datas);
    }

    /**
     * Success 201 : Created.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseCreated(string $message, array $datas = [])
    {
        return $this->successResponse(201, $message, $datas);
    }

    /**
     * Success 202 : Accepted.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseAccepted(string $message, array $datas = [])
    {
        return $this->successResponse(202, $message, $datas);
    }

    /**
     * Error 400 : Bad Request.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseBadRequest(string $message, array $details = [])
    {
        return $this->errorResponse(400, $message, $details);
    }

    /**
     * Error 401 : Unauthorized.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseUnauthorized(string $message, array $details = [])
    {
        return $this->errorResponse(401, $message, $details);
    }

    /**
     * Error 403 : Forbidden.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseForbidden(string $message, array $details = [])
    {
        return $this->errorResponse(403, $message, $details);
    }

    /**
     * Error 404 : Not Found
     *
     * @return \Illuminate\Http\Response
     */
    public function responseNotFound(string $message, array $details = [])
    {
        return $this->errorResponse(404, $message, $details);
    }

    /**
     * Error 405 : Method Not Allowed.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseMethodNotAllowed(string $message, array $details = [])
    {
        return $this->errorResponse(405, $message, $details);
    }

    /**
     * Error 408 : Request Timeout
     *
     * @return \Illuminate\Http\Response
     */
    public function responseRequestTimeout(string $message, array $details = [])
    {
        return $this->errorResponse(408, $message, $details);
    }

    /**
     * Error 429 : Too Many Requests.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseTooManyRequests(string $message, array $details = [])
    {
        return $this->errorResponse(429, $message, $details);
    }

    /**
     * Error 500 : Internal Server Error.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseInternalServerError(string $message, array $details = [])
    {
        return $this->errorResponse(500, $message, $details);
    }

    /**
     * Error 502 : Bad Gateway.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseBadGateway(string $message, array $details = [])
    {
        return $this->errorResponse(502, $message, $details);
    }

    /**
     * Error 503 : Service Unavailable.
     *
     * @return \Illuminate\Http\Response
     */
    public function responseServiceUnavailable(string $message, array $details = [])
    {
        return $this->errorResponse(503, $message, $details);
    }
}


src/ApiResponse.php
<?php

namespace Jornatf\LaravelApiJsonResponse;

use Exception;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Support\Str;
use Jornatf\LaravelApiJsonResponse\Traits\HasJsonResponse;

class ApiResponse
{
    use HasJsonResponse;

    private int $status;

    private string $message;

    private array $datas = [];

    private mixed $details = [];

    private array $validStatus = [
        200 => 'Ok',
        201 => 'Created',
        202 => 'Accepted',
        400 => 'BadRequest',
        401 => 'Unauthorized',
        403 => 'Forbidden',
        404 => 'NotFound',
        405 => 'MethodNotAllowed',
        408 => 'RequestTimeout',
        429 => 'TooManyRequests',
        500 => 'InternalServerError',
        502 => 'BadGateway',
        503 => 'ServiceUnavailable',
    ];

    private int $id;

    /**
     * Json response instance.
     *
     * @return void
     */
    public function __construct()
    {
        //
    }

    /**
     * Method to instance class.
     *
     * @param  int  $status
     * @return ApiResponse
     */
    public function response($status)
    {
        if (! is_int($status) || ! isset($this->validStatus[$status])) {
            throw new Exception('Status code is not valid.');
        }

        $this->status = $status;

        return $this;
    }

    /**
     * Method to add custom message.
     *
     * @param  string  $message
     * @return ApiResponse
     */
    public function addMessage($message)
    {
        if (! is_string($message)) {
            throw new Exception('addMessage(): must contain an array.');
        }

        $this->message = $message;

        return $this;
    }

    /**
     * Method to add datas.
     *
     * @param  array  $datas
     * @return ApiResponse
     */
    public function addDatas($datas)
    {
        if (! is_array($datas)) {
            throw new Exception('addDatas(): must contain an array.');
        }

        $this->datas = $datas;

        return $this;
    }

    /**
     * Method to find a model data.
     *
     * @param  string  $model
     * @param  int  $id
     * @return ApiResponse
     */
    public function find($model, $id)
    {
        $found = $model::find($id);

        $modelName = $this->getModelName($model);

        if (! $found) {
            $this->status = 404;
            $this->addMessage("{$modelName} Not Found");
        } else {
            $this->status = 200;
            $this->addDatas($found->toArray());
            $this->addMessage("{$modelName} Found");
        }

        return $this;
    }

    /**
     * Method to add details.
     *
     * @param  mixed  $details
     * @return ApiResponse
     */
    public function addDetails($details)
    {
        if (! in_array(gettype($details), ['array', 'string'])) {
            throw new Exception('addDetails(): must contain array or a string.');
        }

        if (is_array($this->details)) {
            $this->details[] = $details;
        } else {
            $this->details = $details;
        }

        return $this;
    }

    /**
     * Method to return JSON response.
     *
     * @return ApiResponse
     */
    public function json()
    {
        $method = $this->getMethod();

        $message = $this->message ?? $this->getStatusText();

        return ($this->status < 400)
            ? $this->$method($message, $this->datas)
            : $this->$method($message, $this->details);
    }

    /**
     * Method to get default status text.
     *
     * @return string
     */
    protected function getStatusText()
    {
        return Str::title(Str::snake($this->validStatus[$this->status], ' '));
    }

    /**
     * Returns response method.
     *
     * @return string
     */
    protected function getMethod()
    {
        return 'response'.$this->validStatus[$this->status];
    }

    /**
     * Returns class name without namespace.
     *
     * @param  string  $model
     * @return string
     */
    protected function getModelName($model)
    {
        $model = explode('\\', $model);

        return end($model);
    }
}
